osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -107,17 +107,44 @@ public void AddBeatmap(BeatmapSetInfo beatmapSet)
             });
         }
 
-        public void RemoveBeatmap(BeatmapSetInfo beatmapSet)
+        public void RemoveBeatmap(BeatmapSetInfo beatmapSet) => removeGroup(groups.Find(b => b.BeatmapSet.ID == beatmapSet.ID));
+
+        internal void UpdateBeatmap(BeatmapInfo beatmap)
         {
-            Schedule(delegate
+            // todo: this method should not run more than once for the same BeatmapSetInfo.
+            var set = manager.Refresh(beatmap.BeatmapSet);
+
+            // todo: this method should be smarter as to not recreate panels that haven't changed, etc.
+            var group = groups.Find(b => b.BeatmapSet.ID == set.ID);
+
+            if (group == null)
+                return;
+
+            var newGroup = createGroup(set);
+
+            int i = groups.IndexOf(group);
+            groups.RemoveAt(i);
+            groups.Insert(i, newGroup);
+
+            if (selectedGroup == group && newGroup.BeatmapPanels.Count == 0)
+                selectedGroup = null;
+
+            Filter(null, false);
+
+            //check if we can/need to maintain our current selection.
+            if (selectedGroup == group && newGroup.BeatmapPanels.Count > 0)

```

Comment:
This `&& newGroup.BeatmapPanels.Count > 0` check is needed, since it's covered by the above if.

Reply:
99% sure it is required, since the selection changes in the `Filter` call between the two calls

Reply:
I see.